### PR TITLE
[Messenger] Move PostgreSQL LISTEN/NOTIFY blocking to worker idle event listener

### LINKED ISSUE #46862: [DoctrineMessenger, Postgres] Postgres LISTEN/NOTIFY improvement proposition for better handling of time limit, multi-queue worker and delayed tasks
### Description

Hello,

I'm evaluating a use case where queue workers are running with a 10-minute "get_notify_timeout" option (i.e. the max time the [pgsqlGetNotify()](https://github.com/symfony/doctrine-messenger/blob/9695a3eac87d4fe740d74c6f6ff7912a1cdd5588/Transport/PostgreSqlConnection.php#L82) will synchronously wait for notifications from Postgres), and I found a few problems with it, to which I would like to propose a solution. The problems and solutions are interconnected, so it's perhaps best to discuss them in one ticket and then, if needed, create separate tickets/PRs (which I can do when time comes).

# Problem no. 1 (critical)

Imagine there are two Doctrine queues: `high_priority` and `low_priority`, and a queue worker is started with the following cli command:
```
bin/console messenger:consume high_priority low_priority
```
In other words, the worker will consume messages from both queues, preferring the high_priority. Steps to reproduce the problem:

1. Have no tasks queued for both queues.
2. Start the worker.
    1. Due to an [implementation detail](https://github.com/symfony/doctrine-messenger/blob/9695a3eac87d4fe740d74c6f6ff7912a1cdd5588/Transport/PostgreSqlConnection.php#L62), the following happens:
        1. The worker will conclude that "high_priority" queue is empty and make a note of it. It will not sync-wait on postgres' notifications at this point yet.
        2. The worker will conclude that "low_priority" queue is empty and make a note of it.
        3. The worker enters a new "[master while](https://github.com/symfony/messenger/blob/315d93498bc11c22ccf9ba9fa436380488b8fb9a/Worker.php#L94)" loop.
        4. While attempting to get a new queue task for the "high_priority" queue, the worker executes a blocking sync-wait for a postgres notification [here](https://github.com/symfony/doctrine-messenger/blob/9695a3eac87d4fe740d74c6f6ff7912a1cdd5588/Transport/PostgreSqlConnection.php#L82)
    2. It's important to let the worker block at the sync-wait before continuing with these "steps to reproduced". In practice, one should wait for about a second.
3. Add a new task for the "low_priority" queue. One can do it by hand via an INSERT sql statement with appropriate data.
4. Postgres delivers a notification to the worker that a new record has been added to the messenger_queue db table.
5. The worker "wakes up" from the sync-wait.
6. The worker concludes that the new record is not for the "high_priority" queue and continues to the next queue.
7. The worker attempts to retrieve a task for the "low_priority" queue. Due to the fact that [the "queueEmptiedAt" hasn't been cleared for the "low_priority" queue connection](https://github.com/symfony/doctrine-messenger/blob/9695a3eac87d4fe740d74c6f6ff7912a1cdd5588/Transport/PostgreSqlConnection.php#L62), the worker enters a blocking sync-wait on the "low_priority" queue.

In the last step, the worker should have retrieved the new task for the "low_priority" queue, but instead it entered a blocking sync-wait. This is a problem, especially when the timeout for the sync-wait is 10 minutes, as I mentioned before.

# Problem no. 2

The worker's "--time-limit" option doesn't work as expected, when the worker is in the process of sync-waiting for new queue tasks.

# Problem no. 3

In case there is a delayed task queued that should be picked up for processing after the next 1 minute passes, and the worker enters a sync-wait with a timeout of 10 minutes, then the delayed task will be processed only after the worker times out the sync-wait (assuming nothing else wakes up the worker from the sync-wait).

# Improvement proposition no. 1

I gave these problems a few thoughts and would like to propose the following changes/improvements.

The sync-wait should not happen in [PostgreSqlConnection::get()](https://github.com/symfony/doctrine-messenger/blob/9695a3eac87d4fe740d74c6f6ff7912a1cdd5588/Transport/PostgreSqlConnection.php#L60). Instead, it should happen during the [WorkerRunningEvent](https://github.com/symfony/messenger/blob/315d93498bc11c22ccf9ba9fa436380488b8fb9a/Worker.php#L128). In other words, conceptually, **the worker should not sync-wait while attempting to retrieve new tasks, but instead should sync-wait while being idle** (i.e. in the [idle sleep](https://github.com/symfony/messenger/blob/315d93498bc11c22ccf9ba9fa436380488b8fb9a/Worker.php#L127) if-block). In further other words, **instead of the sync-wait attempting to answer the question of "has potentially a new task been added for queue X", it should answer the question of "has something interesting happen on the messenger_queue db table"**.

This would effectively move the majority of the [PostgreSqlConnection::get()](https://github.com/symfony/doctrine-messenger/blob/9695a3eac87d4fe740d74c6f6ff7912a1cdd5588/Transport/PostgreSqlConnection.php#L60) code to an event listener, and would allow the `PostgreSqlConnection::get()` to not override the base class implementation. I believe it would also allow for [the idiosyncrasy in the base class](https://github.com/symfony/doctrine-messenger/blob/8881ca77e18b4f7d52083811462e1f3a34ad107b/Transport/Connection.php#L193) to be removed.

This would fix the problem no. 1 mentioned above. It would also allow for an easier fix for problems no. 2 and no. 3 (see the next improvement proposition).

# Improvement proposition no. 2

With the improvement proposition no. 1 in place, it would allow the worker to choose the sync-wait timeout more intelligently. I propose that the sync-wait timeout value for the `pgsqlGetNotify()` to be chosen in the following way:

1. Query the queue db table to see if there is at least one pending queue task for the queues that the worker has been started for. If yes, do not sync-wait at all.
    1. This potentially prevents the worker from being pwned by the [postgres' notify grouping behaviour](https://github.com/symfony/doctrine-messenger/blob/8881ca77e18b4f7d52083811462e1f3a34ad107b/Transport/Connection.php#L193), but on second thought, it might not be needed at all, considering the fact that the worker's [idle-wait if-block](https://github.com/symfony/messenger/blob/315d93498bc11c22ccf9ba9fa436380488b8fb9a/Worker.php#L127) isn't executed when there was at least 1 task processed in the past while-loop.
2. Query the queue db table to see if there are any delayed queue tasks for the queues that the worker has been started for. Pick the one that becomes available for processing soonest.
3. Compute the worker's remaining time, as requested by the `--time-limit` cli command option.
4. Out of the 3 time values (the soonest delayed task, the remaining --time-limit, and the configured `get_notify_timeout` option) choose the lowest one, and use it for the `pgsqlGetNotify()` sync-wait call.

This fixes problems no. 2 and no. 3. It perhaps is also simpler conceptually, as the sync-waiting logic is cleanly isolated in its own symfony event listener, rather than being a part of an overridden method in `PostgreSqlConnection::get()`.

# Closing words

Implementing the sync-waiting as proposed above would also conceptually usher in a solid opinion on how sync-waiting should be done in Messenger (compared to polling), i.e. the sync-waiting transport should be recommended to sync-wait on any change happening on the underlying queue(s), rather than only sync-waiting on one queue's new tasks.

I truly believe that this is the way to move forward with this, and for it to have an immense value for the ecosystem (one value being that --time-limit, multi-queue workers, and delayed queue tasks would JustWorkTM (properly), even if the worker is in the process of sync-waiting). I can prepare the code change if the proposition is accepted.

Thank you.

### Example

No real example, as the ticket proposes an underlying implementation change, rather than dev-facing interface change (although, the change allows for more features in Messenger to work, when sync-waiting).

### PR DESCRIPTION
**Pull request by d-ph**
| Q             | A
| ------------- | ---
| Branch?       | 8.1
| Bug fix?      | yes
| New feature?  | yes
| Deprecations? | no
| Tickets       | Fix #46862
| License       | MIT
| Doc PR        | -

When a Messenger worker consumes from multiple PostgreSQL-backed queues, the existing `PostgreSqlConnection::get()` blocks on `pgsqlGetNotify()` for one transport. This prevents the other transports from being polled, effectively breaking priority-based multi-queue consumption.

This PR moves the LISTEN/NOTIFY blocking from `PostgreSqlConnection::get()` into a new `PostgreSqlNotifyOnIdleListener` event subscriber. The listener hooks into the worker lifecycle:

- **`WorkerStartedEvent`**: registers `LISTEN` on all PostgreSQL connections used by the worker, so no notifications are missed from the very first loop iteration.
- **`WorkerRunningEvent`** (idle): blocks on `waitForNotify()` using the first registered connection, replacing the worker's default polling sleep with efficient LISTEN/NOTIFY wake-up.

This way, the worker's main loop can check all transports in priority order on every iteration, and only blocks *after* all queues have been found empty.

### Backwards compatibility

- `DoctrineTransportFactory` accepts an optional `?PostgreSqlNotifyOnIdleListener` second constructor argument (defaults to `null`) — fully BC.
- When no listener is wired (e.g. standalone usage without the bundle), `PostgreSqlConnection::get()` preserves the original blocking LISTEN/NOTIFY fallback behavior — no functionality is lost.
- When `listen()` is called externally (by the listener), `get()` delegates straight to `parent::get()` without blocking, letting the listener handle the idle wait instead.
- No restrictions on mixing PostgreSQL and non-PostgreSQL transports in the same worker.

### Changes

- **`PostgreSqlConnection`**: added public `listen()` and `waitForNotify()` methods; `get()` retained as fallback when no external listener is active.
- **`PostgreSqlNotifyOnIdleListener`** (new): `EventSubscriberInterface` that registers LISTEN on start and blocks on NOTIFY when idle.
- **`DoctrineTransportFactory`**: optional listener injection; registers connections when a `transport_name` is provided.

### CODE REVIEW

**Comment by christian-kolb** [comment]
@nicolas-grekas Even though this is includes a logic change, shouldn't it be considered a bug fix as at multiple queues (which just means prioritizing messages) is kind of broken at the moment?
@d-ph I'm having the same issue in my project and would love to use this fix. Any way I can support you on it?

**Comment by d-ph** [comment]
Hi Christian,

> Any way I can support you on it?

I'm waiting for someone to say "yes/ok" to the points: 2, 3 (to allow an addition of the getter), 6 (to include the "additional" improvement in the scope of this PR), and 7. When I get an approval (or at least an interest in principle), then I'll go ahead and complete this PR (but I might have some further questions about unit testing, and online documentation). So if it's in your ability to obtain this, then please do.

The PR contains a running and working code (but note the "sibling" PR for DoctrineBundle), however at present it only includes the code for problem 1, mentioned in #46862. Feel free to fork the 2 git repositories, and install them via composer in your app. The code is obviously not unit-tested, so it may break between Messenger minor/major updates, but it does work at present.

Not much else I can say in that regard. At the very least, I'm not going to invest my time in finishing this PR, if I don't see any interest from the Symfony maintainers. I might at some point write a separate Postgres(Messenger)Transport with the changes mentioned in the github ticket, just for my own use.

**Comment by christian-kolb** [comment]
@d-ph Unfortunately that's not going to be me as I'm not part of that group. But @nicolas-grekas marked it for the next milestone. Isn't that kind of an confirmation?

**Comment by d-ph** [comment]
Chris,

I'm not going to play a "what the poet meant" regarding milestones assignment.

I don't think I ask for much, considering the fact that I'm going to be doing the heavy-lifting.

**Comment by christian-kolb** [comment]
I'm sorry. That wasn't meant as a push but rather as "I just realized, did you see that". I also don't know what it means ether, and hoped you might.
You're definitively not asking for too much.
@nicolas-grekas Can you confirm the scope that is tackled here and that this approach would be merged and supported?

**Comment by bendavies** [comment]
this PR needs to take https://github.com/symfony/symfony/pull/47209 into account

**Comment by christian-kolb** [comment]
@nicolas-grekas This PR was moved to different milestones twice, but the author is still waiting on a confirmation / guidance from you before finalizing the PR. Otherwise I guess this will be moved to the next milestone again 😞

**Comment by christian-kolb** [comment]
@nicolas-grekas @fabpot This PR was moved to different milestones again, but the author is still waiting on a confirmation / guidance from you before finalizing the PR. Otherwise I guess this will be moved to the next milestone again.

**Comment by keichinger** [comment]
We're running into the same issues with PostgreSQL's `LISTEN/NOTIFY` feature not correctly working when starting a message consumer with multiple queues. It works as expected with a single queue name, though we'd prefer for sake of simplicity to keep a single message worker instead of N workers.

@nicolas-grekas @fabpot could one of you please have a look at @d-ph's questions from the initial comment, so we could hopefully ( 🤞 ) get this PR moving forward? :)

**Comment by nicolas-grekas** [comment]
I patched these files recently in #62990 and #62988
This means this PR needs a rebase, and also maybe checking if the issue still exists after the changes?

**Review by nicolas-grekas — APPROVED** [review]
PR now ready, targeting 8.1.
PR description updated.
Sidekick ready also at https://github.com/doctrine/DoctrineBundle/pull/2198
TL;DR, this adds an optional `PostgreSqlNotifyOnIdleListener` (wired by the PR on DoctrineBundle) that takes care of listening on PG for all queues at once when the worker is idle. This remains compatible with non-PG queues.

**Comment by nicolas-grekas** [comment]
Thank you @d-ph.

**Comment by d-ph** [comment]
Hi Nicolas. Thanks for completing this PR, and apologies for not coming back to your last comment. January was busy for me (and February will be also). I will try this Friday to check all the changes that you've made, and see whether there's anything for me to make any additional comment on. I realise that this PR is already merged, so it's not like I would be able to "stop" anything, but I will still spend a moment to understand your changes and see what's what. 

Thank you.

**Comment by nicolas-grekas** [comment]
that'd be great thanks !

**Comment by d-ph** [comment]
Hi,

I managed to find some time to review this. Please see the following comments.

1. [In here](https://github.com/symfony/symfony/blob/b296fa61edd2c13ca9d7306864946890e38453f0/src/Symfony/Component/Messenger/Bridge/Doctrine/EventListener/PostgreSqlNotifyOnIdleListener.php#L55), `PostgreSqlNotifyOnIdleListener.php` calls `PostgreSqlConnection.php::listen()` on _ALL_ involved instances of `PostgreSqlConnection` (i.e. on "worker receivers" that a worker cli command is run for), but then the `PostgreSqlConnection.php::waitForNotify()` is called only for the first one [here](https://github.com/symfony/symfony/blob/b296fa61edd2c13ca9d7306864946890e38453f0/src/Symfony/Component/Messenger/Bridge/Doctrine/EventListener/PostgreSqlNotifyOnIdleListener.php#L75). As far as I understand it, calling postgres' `LISTEN`, and then NOT retrieving those postgres' notifications will lead to postgres' running out of the "shared memory" that's used for queuing pg notifications for the pg connections that "subscribed" to them via `LISTEN`. When it happens, I believe postgres would start refusing any connection to dispatch new pg notfications until that "backlog" starts clearing. By "refusing", I mean "refusing to commit db transactions" (which would lead to Doctrine "flush()" crashes). Gpt says that the literal pg error in that case would spell the following: `ERROR:  too many notifications in the NOTIFY queue`.

The reason it's somewhat ok in this PR, is because in order for this to happen, the symfony end user would have to deliberately use multiple Doctrine DBAL connections for their Messenger queues (i.e. transports). The likelihood of this is low. However, when they do do it, they would immediately run into the problem explained above.

The actual problem here is that that PR uses `PostgreSqlConnection.php::listen()` to do TWO things: a) execute the `LISTEN` sql, and b) set `::notifyHandledExternally` to `true` in order to force `parent::get()` to be called in [here](https://github.com/symfony/symfony/blob/b296fa61edd2c13ca9d7306864946890e38453f0/src/Symfony/Component/Messenger/Bridge/Doctrine/Transport/PostgreSqlConnection.php#L65). We need "b)" to happen for all `PostgreSqlConnection`s, but for "a)" to happen only on the "first" `PostgreSqlConnection`. Just note that my original PR did just that (although it also removed the entire `PostgreSqlConnection::get()` override, which caused a BC break).

Ideally, I would prefer that `LISTEN` is not called on any other `PostgreSqlConnection` but the one that then gets the `waitForNotify()` call. Perhaps this could be achieved by adding an "option argument" to `PostgreSqlConnection::listen()` which would allow `PostgreSqlNotifyOnIdleListener.php` to "achieve b) without a)" for all subsequent instances of `PostgreSqlConnection`.

I will leave this decision to you.

2. The PR says `No restrictions on mixing PostgreSQL and non-PostgreSQL transports in the same worker`, but there is nothing in the PR that prevents "nonsense" from happening. I.e. a worker can be started for a mixture of PG and non-PG transports but if it decides to "go to a postgres notify sleep", it will prevent non-PG transports from being processed. So I don't know what the decision is here because in this PR the worker can be started for a mixture of PG/non-PG queues but it will not work for the non-PG queues. Is there a plan to "allow it to run like this but mention that it does not work in the user documentation"?

3. In my original PR, I was requiring that all involved `PostgreSqlConnection`s have a "minimal common config" in [here](https://github.com/d-ph/symfony/blob/6d9e15abcdd9fca8bd5cab2f532bd13e13e03760/src/Symfony/Component/Messenger/Bridge/Doctrine/EventListener/PostgreSqlWaitForPgNotifyOnIdleListener.php#L122). Namely, that all `PostgreSqlConnection`s have the following options exactly the same: `connection`, `table_name`, `get_notify_timeout`. The PR that got merged, did not include that code. Please note that without that code, it will lead to "silent failures" when for some reason the aforementioned config is different. The failures might be minor (e.g. the worker pg-sleeps for a different amount of time than configured), or major (e.g. when the `PostgreSqlConnection`s use different table names or DBAL connections, leading to "oversleeping" (i.e. lack of queue processing)).

I would prefer if that "config check" was included in this PR but I will leave this decision to you.

4. I don't exactly like how this PR makes use of the `check_delayed_interval` pg transport option in [here](https://github.com/symfony/symfony/blob/b296fa61edd2c13ca9d7306864946890e38453f0/src/Symfony/Component/Messenger/Bridge/Doctrine/EventListener/PostgreSqlNotifyOnIdleListener.php#L69). I generally consider the `check_delayed_interval` option as redundant, once "improvement proposition number 2" is implemented from the related Github ticket. As far as I can see, the only reason that option is used in this PR, is because it has a sensible default value, while `get_notify_timeout` does not have a sensible default value (i.e. it's set to 0 by default, effectively making the worker never "pg-sleep" (or sleep without a timeout, depending on which part of the code is looked at)). In my opinion, the problem is the not ideal default value of `get_notify_timeout`, instead of anything else.

Also please note how `check_delayed_interval` is used in [PostgreSqlConnection.php::get()](https://github.com/symfony/symfony/blob/b296fa61edd2c13ca9d7306864946890e38453f0/src/Symfony/Component/Messenger/Bridge/Doctrine/Transport/PostgreSqlConnection.php#L80). It actually _computes_ a timeout value based on when the queue was last time emptied. I.e. the _meaning_ of the `check_delayed_interval` option is different there, than it is in this PR. In my opinion, this is another reason NOT to use it in this PR.

I would prefer if that "option" was not used in this PR, and that instead the default value of `get_notify_timeout` is set to something sensible, perhaps only if the worker uses "pg LISTEN/NOTIFY"

I'll leave it to you.

5. Lastly, it's perhaps worthwhile mentioning that the current PR implements only the "improvement number 1" from the related Github ticket, without the "improvement number 2". Note tha without that second improvement, the "pg sync-sleeping" is still susceptible to problems number 2 and 3, described in the related ticket. I.e. the `--time-limit` Worker cli argument, and the delayed queue tasks will continue to be "overslept".

I would say that the "improvement number 1" was by far the bigger challenge in the related ticket. May I propose that you allocate the extra 1-2 days to complete this PR with what the related ticket mentioned in "improvement number 2", please (because I assume that the surroundings of this code are still "fresh" in your mind)? This way, this entire "feature" would be _completely_ fixed in Symfony 8.1, and it would truly put this matter to rest.

In other words, the "pg timeout" chosen [here](https://github.com/symfony/symfony/blob/b296fa61edd2c13ca9d7306864946890e38453f0/src/Symfony/Component/Messenger/Bridge/Doctrine/EventListener/PostgreSqlNotifyOnIdleListener.php#L69) has to actually be chosen more "intelligently" (i.e. as described in the related Github ticket). 

---

Other than that, I have to admit that the current PR does fix the "problem number 1" in the related Github ticket, so that's absolutely job well done. Thanks.

Please let me know if I can assist in anything else relating to this PR.

Regards,

**Comment by nicolas-grekas** [comment]
Thanks for the nice review, see https://github.com/symfony/symfony/pull/63402
